### PR TITLE
Composite type validations and DML representation

### PR DESCRIPTION
**Pull request by tomhoule**
- [x] Validation in ParserDatabase
- [x] Walkers in ParserDatabase
- [x] Representation in dml
- [x] Lifting to dml

For the design of validations, see https://www.notion.so/prismaio/Supporting-Embedded-Documents-in-MongoDB-4a8a04cd046c4bdf9665008df9c6288c?d=d920dfae48794dd5a5badb165117fd48#29cbfa1fd8744170b331c3e07c9df71a

Left to do before we can close https://github.com/prisma/prisma/issues/8603 : only `@map`.

### CODE REVIEW

File: query-engine/prisma-models/src/datamodel_converter.rs [hunk 1]
```diff
@@ -445,6 +445,7 @@ trait DatamodelFieldExtensions {
 impl DatamodelFieldExtensions for dml::ScalarField {
     fn type_identifier(&self) -> TypeIdentifier {
         match &self.field_type {
+            dml::FieldType::CompositeType(_) => todo!("composite type support in datamodel_converter"),
```
Comment (tomhoule) [inline query-engine/prisma-models/src/datamodel_converter.rs:448]: Hand-over to client team happens here.

File: libs/datamodel/core/src/transform/ast_to_dml/db/walkers.rs [hunk 2]
```diff
@@ -13,14 +13,67 @@ use crate::{
...
+    db: &'db ParserDatabase<'ast>,
+}
+
+impl<'ast, 'db> CompositeTypeWalker<'ast, 'db> {
+    pub(crate) fn name(&self) -> &'ast str {
+        &self.db.ast[self.ctid].name.name
+    }
+
+    pub(crate) fn fields(&self) -> impl Iterator<Item = CompositeTypeFieldWalker<'ast, 'db>> {
+        let db = self.db;
```
Comment (pimeys) [inline libs/datamodel/core/src/transform/ast_to_dml/db/walkers.rs:41]: You can skip this by adding `+ 'db` to the return type.

File: libs/datamodel/core/tests/base/duplicates.rs [hunk 3]
```diff
@@ -81,6 +81,73 @@ fn fail_on_duplicate_models_with_relations() {
...
+        [1;94m14 | [0m
+        [1;94m15 | [0mtype [1;91mAddress[0m {
+        [1;94m   | [0m
+    "#]];
+
+    expectation.assert_eq(&parse_and_render_error(dml));
+}
+
+#[test]
+fn fail_on_composite_type_model_conflict() {
```
Comment (pimeys) [inline libs/datamodel/core/tests/base/duplicates.rs:119]: Maybe add `fail_on_composite_type_enum_conflict()` test?

File: libs/datamodel/core/tests/capabilities/sqlite.rs [hunk 4]
```diff
@@ -240,3 +240,32 @@ fn key_order_enforcement_support() {
...
+    "#;
+
+    let err = datamodel::parse_schema(schema).unwrap_err();
+
+    let expected = expect![[r#"
+        [1;91merror[0m: [1mError validating: Composite types are not supported on sqlite.[0m
+          [1;94m-->[0m  [4mschema.prisma:7[0m
+        [1;94m   | [0m
+        [1;94m 6 | [0m
+        [1;94m 7 | [0m        [1;91mtype Address {[0m
```
Comment (pimeys) [inline libs/datamodel/core/tests/capabilities/sqlite.rs:264]: No indoc makes these errors a bit ugly...

**Review by pimeys — APPROVED** [review]

**Pull request merged**